// @n8n/expression-runtime/dist/cjs/runtime/reset.js
(function (factory) {
    if (typeof module === "object" && typeof module.exports === "object") {
        var v = factory(require, exports);
        if (v !== undefined) module.exports = v;
    }
    else if (typeof define === "function" && define.amd) {
        define(["require", "exports", "luxon", "../extensions/extend", "../extensions/function-extensions", "./safe-globals", "./lazy-proxy"], factory);
    }
})(function (require, exports) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    exports.resetDataProxies = resetDataProxies;
    const luxon_1 = require("luxon");
    const extend_1 = require("../extensions/extend");
    const function_extensions_1 = require("../extensions/function-extensions");
    const safe_globals_1 = require("./safe-globals");
    const lazy_proxy_1 = require("./lazy-proxy");
    // ============================================================================
    // Reset Function for Data Proxies
    // ============================================================================
    /**
     * Reset workflow data proxies before each evaluation.
     *
     * This function is called from the bridge before executing each expression
     * to clear proxy caches and initialize fresh workflow data references.
     *
     * Pattern:
     * 1. Create lazy proxies for complex properties ($json, $binary, etc.)
     * 2. Fetch primitives directly ($runIndex, $itemIndex)
     * 3. Create function wrappers for callable properties ($items, etc.)
     * 4. Expose all properties to globalThis for expression access
     *
     * Called from bridge: context.evalSync('resetDataProxies()')
     */
    function resetDataProxies(timezone) {
        if (timezone && !luxon_1.IANAZone.isValidZone(timezone)) {
            throw new Error(`Invalid timezone: "${timezone}"`);
        }
        luxon_1.Settings.defaultZone = timezone ?? 'system';
        // Clear existing __data object
        globalThis.__data = {};
        // __sanitize must be on __data because PrototypeSanitizer generates:
        // obj[this.__sanitize(expr)] where 'this' is __data (via .call(__data) wrapping)
        globalThis.__data.__sanitize = safe_globals_1.__sanitize;
        // Verify callbacks are available
        // Note: ivm.Reference may not be typeof 'function', check for existence
        if (!globalThis.__getValueAtPath) {
            throw new Error('__getValueAtPath callback not registered');
        }
        // -------------------------------------------------------------------------
        // Create lazy proxies for complex workflow properties
        // -------------------------------------------------------------------------
        globalThis.__data.$json = (0, lazy_proxy_1.createDeepLazyProxy)(['$json']);
        globalThis.__data.$binary = (0, lazy_proxy_1.createDeepLazyProxy)(['$binary']);
        globalThis.__data.$input = (0, lazy_proxy_1.createDeepLazyProxy)(['$input']);
        globalThis.__data.$node = (0, lazy_proxy_1.createDeepLazyProxy)(['$node']);
        globalThis.__data.$parameter = (0, lazy_proxy_1.createDeepLazyProxy)(['$parameter']);
        globalThis.__data.$workflow = (0, lazy_proxy_1.createDeepLazyProxy)(['$workflow']);
        globalThis.__data.$prevNode = (0, lazy_proxy_1.createDeepLazyProxy)(['$prevNode']);
        globalThis.__data.$data = (0, lazy_proxy_1.createDeepLazyProxy)(['$data']);
        globalThis.__data.$env = (0, lazy_proxy_1.createDeepLazyProxy)(['$env']);
        // -------------------------------------------------------------------------
        // Create DateTime values inside the isolate (not lazy-loaded from host,
        // because host-side DateTime objects lose their prototype crossing the
        // boundary). The isolate has its own luxon with the correct timezone
        // already set via Settings.defaultZone above.
        // -------------------------------------------------------------------------
        globalThis.__data.$now = luxon_1.DateTime.now();
        globalThis.__data.$today = luxon_1.DateTime.now().set({ hour: 0, minute: 0, second: 0, millisecond: 0 });
        // -------------------------------------------------------------------------
        // Fetch primitives directly (no lazy loading needed for simple values)
        // -------------------------------------------------------------------------
        try {
            globalThis.__data.$runIndex = globalThis.__getValueAtPath.applySync(null, [['$runIndex']], {
                arguments: { copy: true },
                result: { copy: true },
            });
        }
        catch (error) {
            // Property doesn't exist - set to undefined
            globalThis.__data.$runIndex = undefined;
        }
        try {
            globalThis.__data.$itemIndex = globalThis.__getValueAtPath.applySync(null, [['$itemIndex']], {
                arguments: { copy: true },
                result: { copy: true },
            });
        }
        catch (error) {
            // Property doesn't exist - set to undefined
            globalThis.__data.$itemIndex = undefined;
        }
        // -------------------------------------------------------------------------
        // Expose workflow data to globalThis for expression access
        // -------------------------------------------------------------------------
        globalThis.$json = globalThis.__data.$json;
        globalThis.$binary = globalThis.__data.$binary;
        globalThis.$input = globalThis.__data.$input;
        globalThis.$node = globalThis.__data.$node;
        globalThis.$parameter = globalThis.__data.$parameter;
        globalThis.$workflow = globalThis.__data.$workflow;
        globalThis.$prevNode = globalThis.__data.$prevNode;
        globalThis.$runIndex = globalThis.__data.$runIndex;
        globalThis.$itemIndex = globalThis.__data.$itemIndex;
        globalThis.$data = globalThis.__data.$data;
        globalThis.$env = globalThis.__data.$env;
        globalThis.$now = globalThis.__data.$now;
        globalThis.$today = globalThis.__data.$today;
        // Expose standalone functions (min, max, average, numberList, zip, $ifEmpty, etc.)
        Object.assign(globalThis.__data, function_extensions_1.extendedFunctions);
        // -------------------------------------------------------------------------
        // Handle function properties (check if value is function metadata)
        // -------------------------------------------------------------------------
        // Check if $items exists and is a function
        if (globalThis.__callFunctionAtPath) {
            try {
                const itemsValue = globalThis.__getValueAtPath.applySync(null, [['$items']], {
                    arguments: { copy: true },
                    result: { copy: true },
                });
                // If it's function metadata, create wrapper
                if (itemsValue && typeof itemsValue === 'object' && itemsValue.__isFunction) {
                    globalThis.$items = function (...args) {
                        return globalThis.__callFunctionAtPath.applySync(null, [['$items'], ...args], {
                            arguments: { copy: true },
                            result: { copy: true },
                        });
                    };
                    globalThis.__data.$items = globalThis.$items;
                }
                else {
                    // Not a function - set to undefined or the value itself
                    globalThis.$items = itemsValue;
                    globalThis.__data.$items = itemsValue;
                }
            }
            catch (error) {
                // Property doesn't exist
                globalThis.$items = undefined;
                globalThis.__data.$items = undefined;
            }
        }
        // -------------------------------------------------------------------------
        // Expose globals on __data so tournament's "x in this ? this.x : global.x"
        // pattern resolves them correctly (tournament checks __data before global)
        // -------------------------------------------------------------------------
        globalThis.__data.DateTime = globalThis.DateTime;
        globalThis.__data.Duration = globalThis.Duration;
        globalThis.__data.Interval = globalThis.Interval;
        // Expose extend/extendOptional on __data so tournament's "x in this ? this.x : global.x"
        // pattern resolves them correctly when the VM checks __data first
        globalThis.__data.extend = extend_1.extend;
        globalThis.__data.extendOptional = extend_1.extendOptional;
        // TODO: Add other function properties as needed ($item, $vars, etc.)
    }
});
